### PR TITLE
compute precise language information with go-enry for lang: queries

### PR DESCRIPTION
**Pull request by rmmh**
This will disambiguate files with similar language extensions, and otherwise understands far more filetypes than the previous ctags-based matching.

### CODE REVIEW

File: matchtree.go
Comment (keegancsmith) [inline matchtree.go]: by directly creating the regexpMatchTree you don't get to take advantage of the trigram indexes. Is there not a way we can re-use that smarts we have here? eg return `d.newMatchTree(&query.Regexp{...})`.

How about this, in `indexData.simplify` you couple this logic by always returning true. Why not generate the equivalent `query.Regexp` in that bit of code and then you don't need to touch matchtree generation. The benefit of this is there is less coupling between simplify and newMatchTree

File: indexbuilder.go
Comment (keegancsmith) [inline indexbuilder.go:195]: I don't really understand why this isn't just uint16? You don't need backwards compatibility here since you just write this out and do not read.

**Review by keegancsmith — COMMENTED** [review]
Looks like backwards compatibility tests are failing on the latest commit for always doing 16-bit.

File: indexbuilder.go
Comment (rmmh) [reply indexbuilder.go:195]: Leaving it uint8 lets it be mmaped instead of converted from a disk format.

File: matchtree.go
Comment (rmmh) [reply matchtree.go]: done

File: eval.go
Comment (keegancsmith) [inline eval.go:129]: can you add a test for this / manually test this? For example this looks like we end up erroring out/panicing, you are returning a MatchTree not a query.

Do we need to worry about case sensitivity?

File: matchtree.go
Comment (keegancsmith) [inline matchtree.go]: if I am not mistaken this is equivalent to the old code? so we don't actually need any changes here?

File: indexbuilder.go
Comment (keegancsmith) [reply indexbuilder.go:195]: do you mean for when reading the older disk formats? Or you just wanna be explicit in the little endian encoding? This array is purely for the encoding, somewhere else we actually do the marshalling. The indexData.languages is still a []byte which I understand since that is how we get data back from our mmap thing.

**Review by keegancsmith — APPROVED** [review]
LGTM, approving to unblock but there is an inline blocking issue that needs to be resolved before landing.

Should we not keep one of our older `testdata/golden/TestReadSearch/*.golden` fixtures so we can test the backwards compatibility? For example I picked up a bug now in the fallback lang logic, and the older shards would be perfect for testing against.

File: matchtree.go
Comment (rmmh) [reply matchtree.go]: fixed

File: eval.go
Comment (rmmh) [reply eval.go:129]: added a test and fixed this, it's case insensitive

**Comment by rmmh** [comment]
@keegancsmith let's address where the languages marshaling happens in a future change

**Review by keegancsmith — COMMENTED** [review]
LGTM

**Pull request merged**
